File: server/common/protocol.py
from .utils import Bet


BET_PKT = 1
BET_ACK_PKT = 2

class PacketHeader:
    """Header common to every packet in the protocol"""
    def __init__(self,pkt_type,id):
        self.packet_type = pkt_type
        self.id = id
        
    def header_to_bytes(self,payload_size):
        return [self.packet_type,self.id,payload_size + 3]
        

class BetPacket:
    """A packet containing information of a bet"""
    def __init__(self,bet):
        self.pkt_type = BET_PKT
        self.bet = bet
        
def bet_from_bytes(bytes):
    """Converts an array of bytes into a BetPacket"""
    msg = ''.join(chr(i) for i in bytes[3:])
    agency = str(bytes[1])
    fields = msg.split("|")
    bet = Bet(agency,fields[0],fields[1],fields[2],fields[3],fields[4])
    return BetPacket(bet)


class BetAckPacket:
    """Bet Ack packet sent to the client"""
    def __init__(self,document,number,status,id):
        self.header = PacketHeader(BET_ACK_PKT,id)
        self.document = document
        self.number = number
        self.status = status
        
    def bet_ack_to_bytes(self):
        """Converts a BetAckPacket to an array of bytes"""
        format_payload = self.document + "|" + self.number + "|" + self.status
        header_bytes = self.header.header_to_bytes(len(format_payload))
        return bytearray(header_bytes) + format_payload.encode('utf-8')
         

File: client/common/protocol.go
package common

import (
	"strings"
)

// Packet Header common to every packet in the protocol
type PacketHeader struct {
	PacketType uint8
	ID         uint8
}

// Converts PacketHeader to bytes
func (h *PacketHeader) HeaderToBytes(PayloadSize uint8) []byte {
	bytes := []byte{h.PacketType, h.ID, PayloadSize + 3}
	return bytes
}

// Data from a bet
type BetData struct {
	Name     string
	Surname  string
	Document string
	Birthday string
	Number   string
}

// Bet packet
type Bet struct {
	Header PacketHeader
	Data   BetData
}

// Returns a new Bet Packet
func NewBet(data BetData, id uint8) *Bet {
	bet := &Bet{
		Header: PacketHeader{
			PacketType: 1,
			ID:         id,
		},
		Data: data,
	}
	return bet
}

// Converts a Bet struct into an array of bytes
func (b *Bet) BetToBytes() []byte {
	formated_string := b.Data.Name + "|" + b.Data.Surname + "|" + b.Data.Document + "|" + b.Data.Birthday + "|" + b.Data.Number
	header_bytes := b.Header.HeaderToBytes(uint8(len(formated_string)))
	payload_bytes := []byte(formated_string)
	return append(header_bytes, payload_bytes...)
}

// Data from a BetAck Packet sent from the server
type BetAckData struct {
	Document string
	Number   string
	Status   string
}

// BetAck packet
type BetAck struct {
	Header PacketHeader
	Data   BetAckData
}

// Converts an array of bytes into a BetAckPacket
func BetAckFromBytes(bytes []byte) *BetAck {
	data := string(bytes[3:])
	fields := strings.Split(data, "|")
	betAck := &BetAck{
		Header: PacketHeader{
			PacketType: bytes[0],
			ID:         bytes[1],
		},
		Data: BetAckData{
			Document: fields[0],
			Number:   fields[1],
			Status:   fields[2],
		},
	}
	return betAck
}
